programs/fileio_asyncio.c

```diff
@@ -363,3 +365,223 @@ void AIO_WritePool_free(WritePoolCtx_t* ctx) {
+    if(file != NULL)
+        AIO_ReadPool_startReading(ctx);
+}
+
+/* AIO_ReadPool_create:
+ * Allocates and sets and a new readPool including its included jobs.
+ * bufferSize should be set to the maximal buffer we want to read at a time, will also be used
+ * as our basic read size. */
+ReadPoolCtx_t* AIO_ReadPool_create(FIO_prefs_t* const prefs, size_t bufferSize) {
+    ReadPoolCtx_t* ctx = (ReadPoolCtx_t*) malloc(sizeof(ReadPoolCtx_t));
+    if(!ctx) EXM_THROW(100, "Allocation error : not enough memory");
+    AIO_IOPool_init(&ctx->base, prefs, AIO_ReadPool_executeReadJob, bufferSize);
+
+    ctx->srcBufferBaseSize = 2 * bufferSize;
+    ctx->srcBufferBase = (U8*) malloc(ctx->srcBufferBaseSize);
+    ctx->srcBuffer = ctx->srcBufferBase;
+    ctx->srcBufferLoaded = 0;
+    ctx->completedJobsCount = 0;
+
+    if(ctx->base.threadPool)
+        if (ZSTD_pthread_cond_init(&ctx->jobCompletedCond, NULL))
+            EXM_THROW(103,"Failed creating write jobCompletedCond mutex");
+
+    return ctx;
+}
+
+/* AIO_ReadPool_free:
+ * Frees and releases a readPool and its resources. Closes source file. */
+void AIO_ReadPool_free(ReadPoolCtx_t* ctx) {
+    if(AIO_ReadPool_getFile(ctx))
+        AIO_ReadPool_closeFile(ctx);
+    if(ctx->base.threadPool)
+        ZSTD_pthread_cond_destroy(&ctx->jobCompletedCond);
+    AIO_IOPool_destroy(&ctx->base);
+    free(ctx->srcBufferBase);
+    free(ctx);
+}
+
+/* AIO_ReadPool_consumeBytes:
+ * Consumes byes from srcBuffer's beginning and updates srcBufferLoaded accordingly. */
+void AIO_ReadPool_consumeBytes(ReadPoolCtx_t *ctx, size_t n) {
+    assert(n <= ctx->srcBufferLoaded);
+    assert(ctx->srcBuffer + n <= ctx->srcBufferBase + ctx->srcBufferBaseSize);
+    ctx->srcBufferLoaded -= n;
+    ctx->srcBuffer += n;
+}
+
+/* AIO_ReadPool_fillBuffer:
+ * Makes sure buffer has at least n bytes loaded (as long as n is not bigger than the initalized bufferSize).
+ * Returns if srcBuffer has at least n bytes loaded or if we've reached the end of the file.
+ * Return value is the number of bytes added to the buffer.
+ * Note that srcBuffer might have up to 2 times bufferSize bytes. */
+size_t AIO_ReadPool_fillBuffer(ReadPoolCtx_t *ctx, size_t n) {
+    IOJob_t *job;
+    size_t srcBufferOffsetFromBase;
+    size_t srcBufferRemainingSpace;
+    size_t bytesRead = 0;
+    assert(n <= ctx->srcBufferBaseSize/2);
+    while (ctx->srcBufferLoaded < n) {
+        job = AIO_ReadPool_getNextCompletedJob(ctx);
+        if(job == NULL)
+            break;
+        srcBufferOffsetFromBase = ctx->srcBuffer - ctx->srcBufferBase;
+        srcBufferRemainingSpace = ctx->srcBufferBaseSize - (srcBufferOffsetFromBase + ctx->srcBufferLoaded);
+        if(job->usedBufferSize > srcBufferRemainingSpace) {
+            memmove(ctx->srcBufferBase, ctx->srcBuffer, ctx->srcBufferLoaded);
+            ctx->srcBuffer = ctx->srcBufferBase;
+        }
+        memcpy(ctx->srcBuffer + ctx->srcBufferLoaded, job->buffer, job->usedBufferSize);
```

Comment:
Is there something you can do to avoid these `memcpy()`?

Do they actually happen in the normal case? E.g. if zstd is consuming the entire buffer do we ever hit this case?

If they don't happen for zstd, and only happen for other formats, then it probably isn't a big deal.

Reply:
Two possible ways to avoid `memcpy` (at the cost of code complexity) are: 
1. Continue holding the `IOJob_t` and referencing the buffer it owns. This doesn't remove the need for memcpy completely as there are cases where the code would ask for more bytes than are currently available and so we will need to coalesce two buffers. For that we could have another buffer allocated that will be used only when strickly needed.
2. Implement a ring buffer and have each `IOJob_t` reference a portion of this ring buffer.

The `memcpy/memove` bothered me, but I believe these two solutions increase the complexity of the code without actually improving performance. Memcpy should be cheap compared to our other actions (compression / decompression and IO). Am I wrong to assume memcpy is relatively cheap?

Reply:
> Am I wrong to assume memcpy is relatively cheap?

Relatively cheap, yes. Negligible, no.

Decompression runs at > 1GB/s. So adding an extra memcpy can be a 10% slowdown.

Compression can run at a few hundred MB/s, or GBs/s with multithreading. So an extra memcpy can be a 5% slowdown.

Definitely worth benchmarking.

Could you have jobs+1 buffers, and reserve 1 buffer for reading?  When you're finished reading, swap that buffer with the next jobs buffer?

Reply:
> Decompression runs at > 1GB/s. So adding an extra memcpy can be a 10% slowdown.
The memcpy can happen only for input, not for output which should have a lower rate.
Other than that, up to now each such memcpy/memove was instead an `fread` which is a memcpy at the best case, but might also involve a syscall (depending on buffering). While the memcpy can have a non negligible cost, it is definitely cheaper than the previous implementation.

>  Could you have jobs+1 buffers, and reserve 1 buffer for reading? When you're finished reading, swap that buffer with the next jobs buffer?
As mentioned on our second thread this is problematic as there are some cases the user would `consume` some bytes leaving us with a lower number of bytes available than required in the next read. In such a cases we'd need to take some bytes from the current buffer and joint them with bytes from the next buffer. This can only be possible using `memcpy` or if the buffers are all a part of one continuous ring buffer (which also has edge cases toward the end of the buffer, unless we use some mmap tricks) .

Reply:
In any case, I will give it a try, and we can judge it afterwards.

Reply:
> In such a cases we'd need to take some bytes from the current buffer and joint them with bytes from the next buffer

Not necessarily. You could loosen the contract that you always return the requested # of bytes. All compressors should have the ability to consume the input, even if it is smaller than the expected size.

So in the case where you have leftover bytes, instead of joining with the next job, just return a shorter read. Then when its consumed the tail of the current buffer, grab the next job's buffer.

Reply:
Wouldn't that increase chances of memcpy to internal decompressor buffers though?

Reply:
> Relatively cheap, yes. Negligible, no.

You are right there. Perf shows 1-3.5% time spent on memcpy.
Also interesting is that `ZSTD_(de)compressStream` spends even more time on `memcpy`. I wonder if this is related to the code getting less data than expected.